File: app/src/test/java/com/android/surveysyncengine/sync/NetworkDegradationTest.kt
package com.android.surveysyncengine.sync

import com.android.surveysyncengine.FakeSurveyRepository
import com.android.surveysyncengine.TestSyncEngineFactory
import com.android.surveysyncengine.allSucceedApi
import com.android.surveysyncengine.buildResponse
import com.android.surveysyncengine.noNetworkAtIndexApi
import com.android.surveysyncengine.timeoutAtIndexApi
import com.surveysyncengine.data.remote.api.FailureType
import com.surveysyncengine.data.remote.api.FakeSurveyApiService
import com.surveysyncengine.data.remote.api.failurePlan
import com.surveysyncengine.domain.model.SyncStatus
import com.surveysyncengine.sync.NetworkErrorClassifier
import com.surveysyncengine.sync.SyncResult
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

/**
 * Fine-grained tests for early-termination logic across various classifier
 * threshold configurations and failure patterns.
 */
class NetworkDegradationTest {

    private lateinit var repo: FakeSurveyRepository

    @Before
    fun setUp() {
        repo = FakeSurveyRepository()
    }

    @Test
    fun `threshold of 1 stops at first network failure`() = runTest {
        (0..4).forEach { repo.saveResponse(buildResponse(id = "resp-$it")) }
        val engine = TestSyncEngineFactory.create(
            repo = repo,
            api = timeoutAtIndexApi(2),
            classifier = NetworkErrorClassifier(consecutiveFailureThreshold = 1),
        )

        val result = engine.sync() as SyncResult.EarlyTermination

        assertEquals(2, result.succeeded.size)          // resp-0 and resp-1
        assertEquals(1, result.failedBeforeStop.size)   // resp-2
        assertEquals(2, result.remainingCount)          // resp-3 and resp-4
    }

    @Test
    fun `threshold of 3 allows two consecutive failures before stopping`() = runTest {
        (0..9).forEach { repo.saveResponse(buildResponse(id = "resp-$it")) }
        // Two consecutive timeouts at indices 4 and 5 — should NOT stop (threshold=3)
        // Third timeout at index 6 — should stop
        val engine = TestSyncEngineFactory.create(
            repo = repo,
            api = timeoutAtIndexApi(4, 5, 6),
            classifier = NetworkErrorClassifier(consecutiveFailureThreshold = 3),
        )

        val result = engine.sync() as SyncResult.EarlyTermination

        assertEquals(4, result.succeeded.size)          // resp 0-3
        assertEquals(3, result.failedBeforeStop.size)   // resp 4-6
        assertEquals(3, result.remainingCount)          // resp 7-9
    }

    @Test
    fun `no-network error triggers early stop same as timeout`() = runTest {
        (0..5).forEach { repo.saveResponse(buildResponse(id = "resp-$it")) }
        val engine = TestSyncEngineFactory.create(
            repo = repo,
            api = noNetworkAtIndexApi(1, 2),
            classifier = NetworkErrorClassifier(consecutiveFailureThreshold = 2),
        )

        val result = engine.sync()
        assertTrue(result is SyncResult.EarlyTermination)
    }

    @Test
    fun `remaining responses are not touched after early stop`() = runTest {
        (0..7).forEach { repo.saveResponse(buildResponse(id = "resp-$it")) }
        val engine = TestSyncEngineFactory.create(
            repo = repo,
            api = timeoutAtIndexApi(2, 3),
        )

        engine.sync()

        // Responses 4-7 should be PENDING — not attempted
        (4..7).forEach { i ->
            assertEquals(SyncStatus.PENDING, repo.statusOf("resp-$i"))
        }
        // Response 2 and 3 should be FAILED
        assertEquals(SyncStatus.FAILED, repo.statusOf("resp-2"))
        assertEquals(SyncStatus.FAILED, repo.statusOf("resp-3"))
        // Responses 0 and 1 should be SYNCED
        assertEquals(SyncStatus.SYNCED, repo.statusOf("resp-0"))
        assertEquals(SyncStatus.SYNCED, repo.statusOf("resp-1"))
    }

    @Test
    fun `early termination on second sync after network recovers processes remaining`() = runTest {
        (0..5).forEach { repo.saveResponse(buildResponse(id = "resp-$it")) }

        // First sync: stops after resp-1 and resp-2 timeout
        val engine1 = TestSyncEngineFactory.create(
            repo = repo,
            api = timeoutAtIndexApi(1, 2),
        )
        engine1.sync()

        // Network recovers — second sync should process remaining items
        val engine2 = TestSyncEngineFactory.create(
            repo = repo,
            api = allSucceedApi(),
        )
        val result = engine2.sync() as SyncResult.Completed

        // Should retry resp-1 and resp-2 (FAILED) plus resp-3,4,5 (PENDING)
        assertEquals(5, result.succeeded.size)
        assertTrue(result.failed.isEmpty())
    }

    @Test
    fun `server errors interspersed with network errors do not accelerate abort`() = runTest {
        (0..8).forEach { repo.saveResponse(buildResponse(id = "resp-$it")) }

        // Pattern: network, server, network — not 2 consecutive network
        val engine = TestSyncEngineFactory.create(
            repo = repo,
            api = FakeSurveyApiService(
                failurePlan = failurePlan(
                    1 to FailureType.TIMEOUT,
                    2 to FailureType.SERVER_ERROR_500,
                    3 to FailureType.NO_NETWORK,
                )
            ),
            classifier = NetworkErrorClassifier(consecutiveFailureThreshold = 2),
        )

        // Should NOT abort early — the server error broke the consecutive chain
        val result = engine.sync()
        assertTrue("Expected Completed but got $result", result is SyncResult.Completed)
    }
}

File: ios/SurveySyncEngineTests/Network/NetworkDegradationTest.swift
//
//  NetworkDegradationTests.swift
//  SurveySyncEngineIOS
//

import XCTest

/// Integration tests specifically focused on the Sync Engine's resilience and "fail-fast" behavior.
/// These tests verify that the engine correctly identifies a degraded network environment and
/// terminates the session early to preserve device resources, while ensuring it can resume
/// once conditions improve.
final class NetworkDegradationTests: XCTestCase {

    private var repo: FakeSurveyRepository!

    override func setUp() {
        super.setUp()
        // Using a thread-safe Fake Repository to track status changes during the sync loop
        repo = FakeSurveyRepository()
    }

    /// Verifies that if the threshold is set to 1, the engine immediately stops upon the first network error.
    func test_thresholdOf1_stopsAtFirstNetworkFailure() async throws {
        // GIVEN: 5 pending surveys
        for i in 0...4 {
            try await repo.saveResponse(buildResponse(id: "resp-\(i)"))
        }
        
        // Configure API to timeout on the 3rd item (index 2)
        let engine = TestSyncEngineFactory.create(
            repo: repo,
            api: timeoutAtIndexApi(indices: 2),
            classifier: NetworkErrorClassifier(consecutiveFailureThreshold: 1)
        )

        // WHEN: Sync is executed
        let result = await engine.sync()

        // THEN: The result should be an early termination
        guard case let .earlyTermination(succeeded, failedBeforeStop, _, remainingCount) = result else {
            XCTFail("Expected .earlyTermination but got \(result)")
            return
        }

        XCTAssertEqual(succeeded.count, 2)          // resp-0 and resp-1 succeeded
        XCTAssertEqual(failedBeforeStop.count, 1)   // resp-2 failed and triggered the stop
        XCTAssertEqual(remainingCount, 2)           // resp-3 and resp-4 were never attempted
    }

    /// Verifies that the engine handles multiple consecutive failures correctly based on the threshold.
    func test_thresholdOf3_allowsTwoConsecutiveFailuresBeforeStopping() async throws {
        // GIVEN: 10 pending surveys
        for i in 0...9 {
            try await repo.saveResponse(buildResponse(id: "resp-\(i)"))
        }
        
        // Timeouts at 4, 5, and 6. Threshold is 3, so it should stop AFTER the 3rd failure (index 6).
        let engine = TestSyncEngineFactory.create(
            repo: repo,
            api: timeoutAtIndexApi(indices: 4, 5, 6),
            classifier: NetworkErrorClassifier(consecutiveFailureThreshold: 3)
        )

        // WHEN: Sync is executed
        let result = await engine.sync()

        // THEN: Verify the split between succeeded, failed, and unattempted
        guard case let .earlyTermination(succeeded, failedBeforeStop, _, remainingCount) = result else {
            XCTFail("Expected .earlyTermination but got \(result)")
            return
        }

        XCTAssertEqual(succeeded.count, 4)          // resp 0-3 succeeded
        XCTAssertEqual(failedBeforeStop.count, 3)   // resp 4-6 failed
        XCTAssertEqual(remainingCount, 3)           // resp 7-9 remaining
    }

    /// Ensures that 'No Network' (UnknownHost) errors trigger the same abort logic as timeouts.
    func test_noNetworkError_triggersEarlyStopSameAsTimeout() async throws {
        // GIVEN
        for i in 0...5 {
            try await repo.saveResponse(buildResponse(id: "resp-\(i)"))
        }
        let engine = TestSyncEngineFactory.create(
            repo: repo,
            api: noNetworkAtIndexApi(indices: 1, 2),
            classifier: NetworkErrorClassifier(consecutiveFailureThreshold: 2)
        )

        // WHEN
        let result = await engine.sync()

        // THEN
        if case .earlyTermination = result {
            // Success: NoNetwork treated as a network-level failure
        } else {
            XCTFail("Expected .earlyTermination but got \(result)")
        }
    }

    /// Critical check: Verifies that items following an early termination remain in their original PENDING state.
    func test_remainingResponses_areNotTouchedAfterEarlyStop() async throws {
            // GIVEN: 8 surveys
            for i in 0...7 {
                try await repo.saveResponse(buildResponse(id: "resp-\(i)"))
            }
            
            let engine = TestSyncEngineFactory.create(
                repo: repo,
                api: timeoutAtIndexApi(indices: 2, 3),
                classifier: NetworkErrorClassifier(consecutiveFailureThreshold: 2)
            )

            // WHEN: Sync stops at index 3
            _ = await engine.sync()

            // THEN:
            // 1. Items 4-7 must remain PENDING
            for i in 4...7 {
                let status = try await repo.statusOf(id: "resp-\(i)")
                XCTAssertEqual(status, .pending, "Response \(i) should remain PENDING")
            }
            
            // 2. Items 2 and 3 must be marked FAILED (for retry later)
            let status2 = try await repo.statusOf(id: "resp-2")
            let status3 = try await repo.statusOf(id: "resp-3")
            XCTAssertEqual(status2, .failed)
            XCTAssertEqual(status3, .failed)
            
            // 3. Items 0 and 1 must be SYNCED
            let status0 = try await repo.statusOf(id: "resp-0")
            let status1 = try await repo.statusOf(id: "resp-1")
            XCTAssertEqual(status0, .synced)
            XCTAssertEqual(status1, .synced)
        }
    
    /// Verifies that items skipped during an early termination are processed in a subsequent sync if the network recovers.
    func test_earlyTermination_onSecondSyncAfterNetworkRecovers_processesRemaining() async throws {
        // GIVEN
        for i in 0...5 {
            try await repo.saveResponse(buildResponse(id: "resp-\(i)"))
        }

        // Run 1: Network dies at index 1
        let engine1 = TestSyncEngineFactory.create(
            repo: repo,
            api: timeoutAtIndexApi(indices: 1, 2)
        )
        _ = await engine1.sync()

        // WHEN: Network recovers and a second sync starts
        let engine2 = TestSyncEngineFactory.create(
            repo: repo,
            api: allSucceedApi()
        )
        let result = await engine2.sync()

        // THEN: It should finish the queue
        guard case let .completed(succeeded, failed) = result else {
            XCTFail("Expected .completed but got \(result)")
            return
        }

        // Successfully processed the 5 items that were either FAILED or PENDING from Run 1
        XCTAssertEqual(succeeded.count, 5)
        XCTAssertTrue(failed.isEmpty)
    }

    /// Verifies that server-side errors (500s) do not contribute to a network-down termination.
    /// Logic: If the server returns a 500, the network is working, but the server is struggling.
    func test_serverErrors_interspersedWithNetworkErrors_doNotAccelerateAbort() async throws {
        // GIVEN: 9 surveys
        for i in 0...8 {
            try await repo.saveResponse(buildResponse(id: "resp-\(i)"))
        }

        // Pattern: network timeout -> server error -> no network
        // The server error (index 2) should reset the consecutive failure counter.
        let failurePlan: [Int: FailureType] = [
            1: .timeout,
            2: .serverError500,
            3: .noNetwork
        ]

        let engine = TestSyncEngineFactory.create(
            repo: repo,
            api: FakeSurveyApiService(
                failurePlan: failurePlan.map { index, type in
                    FailureInstruction(index, type)
                }
            ),
            classifier: NetworkErrorClassifier(consecutiveFailureThreshold: 2)
        )

        // WHEN: Sync is executed
        let result = await engine.sync()

        // THEN: It should NOT abort early because there weren't 2 CONSECUTIVE network-level failures.
        if case .completed = result {
            // Success: Processed the whole list (some failed, but loop finished)
        } else {
            XCTFail("Expected .completed (with some failed items) but got \(result)")
        }
    }
}
